### PR TITLE
Add future.v7_relativeSplatPath flag

### PR DESCRIPTION
**Pull request by brophdawg11**
Re-implement the bug fix from #10983 behind a `future.v7_relativeSplatPath` flag because we determined that a lot of folks were relying on the buggy behavior (see #11052 and #11078).

Usage:
```js
/**** Data router ****/
let router = createBrowserRouter(routes, { 
  future: { v7_relativeSplatPath: true }
});
<RouterProvider router={router} />

/**** Non-Data Router ****/
<BrowserRouter future={{ v7_relativeSplatPath: true }}>

/**** Data Router SSR ****/
let { query, dataRoutes } = createStaticHandler(routes, {
  future: { v7_relativeSplatPath: true },
});
let router = createStaticRouter(dataRoutes, context, {
  future: { v7_relativeSplatPath: true },
});
<StaticRouterProvider router={router} context={context} />

/**** Non-Data Router SSR ****/
<StaticRouter future={{ v7_relativeSplatPath: true }}>
```

**Todo:**
* [x] Docs + examples of buggy behavior

### CODE REVIEW

File: packages/router/utils.ts [hunk 1]
```diff
@@ -1145,6 +1145,25 @@ export function getPathContributingMatches<
...
+  T extends AgnosticRouteMatch = AgnosticRouteMatch
+>(matches: T[], v7_relativeSplatPath: boolean) {
+  let pathMatches = getPathContributingMatches(matches);
+
+  // When v7_relativeSplatPath is enabled, use the full pathname for the leaf
+  // match so we include splat values for "." links.  See:
+  // https://github.com/remix-run/react-router/issues/11052#issuecomment-1836589329
+  if (v7_relativeSplatPath) {
+    return pathMatches.map((match, idx) =>
+      idx === matches.length - 1 ? match.pathname : match.pathnameBase
```
Comment (brophdawg11) [inline packages/router/utils.ts:1160]: This is the primary functional code change - when the flag is enabled we leverage `pathname` on the leaf match to ensure we include the splat portion of the path

File: packages/react-router/lib/hooks.tsx [hunk 2]
```diff
@@ -193,12 +193,12 @@ function useNavigateUnstable(): NavigateFunction {
...
 
   let dataRouterContext = React.useContext(DataRouterContext);
-  let { basename, navigator } = React.useContext(NavigationContext);
+  let { basename, future, navigator } = React.useContext(NavigationContext);
   let { matches } = React.useContext(RouteContext);
   let { pathname: locationPathname } = useLocation();
 
   let routePathnamesJson = JSON.stringify(
-    getPathContributingMatches(matches).map((match) => match.pathnameBase)
+    getResolveToMatches(matches, future.v7_relativeSplatPath)
```
Comment (brophdawg11) [inline packages/react-router/lib/hooks.tsx:201]: Everywhere we used to do this `pathnameBase` mapping is now done via `getResolveToMatches`

**Comment by cduff** [comment]
@brophdawg11

I first asked this question over at https://github.com/remix-run/react-router/issues/11052 but got no response, so thought I'd try here.

I was hit by https://github.com/remix-run/react-router/issues/11052 and was just thinking about how I came to rely on this "buggy behavior".

I originally used [Reach Router](https://github.com/reach/router), as recommended by the React Router team at the time (before React Router v6).

I used "Embedded Routers" as per the example on this page: https://reach.tech/router/nesting.

I then migrated to React Router as [recommended by the team](https://reactrouter.com/en/main/upgrading/reach).

Is my understanding correct that the Reach Router "Embedded Routers" example is now considered to have relied on buggy behavior?

Just trying to understand what the future direction might be and how it might impact my app.

Thanks for your time.

**Comment by brophdawg11** [comment]
Yes - the behavior was chosen partially to help support the concept of embedded routers, but did not foresee the other inconsistencies that were introduced by that choice - such that `.` no longer meant the `current location` and path resolution would work inconsistently within `path="dashboard/*"` and `path="dashboard/:page"` etc.  We wrote up some of the inconsistencies in the [changeset](https://github.com/remix-run/react-router/pull/11087/files#diff-4bf55d5252d1b4743b2ae595c19def7dfd1eb5044b255e4cfe58827ad8f022ef) and the [`useResolvedPath` docs](https://reactrouter.com/en/release-next/hooks/use-resolved-path#splat-paths) (currently available on the `release-next` branch)

**Comment by jasikpark** [comment]
Should https://github.com/remix-run/react-router/blob/5567158e1f7ab96683bcb4c757e27b1c22ceb68f/packages/react-router/lib/components.tsx#L61 accept this future flag? 

`return <RouterProvider router={router} future={{ v7_relativeSplatPath: true }} />;` errors for me, since this new tag is filtered out

**Pull request merged**